### PR TITLE
[dbus] dbus-daemon, dbus-launch,... are first class artifacts

### LINKED ISSUE #37625: [dbus] dbus-daemon and dbus-launch are runtime executables and should be provided in both release and debug versions
### Is your feature request related to a problem? Please describe.

dbus-daemon and dbus-launch are only installed in release versions in `${VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/tools/dbus`. This matches the recommendations in [installation-tree-layout#layout-tools](https://learn.microsoft.com/en-us/vcpkg/reference/installation-tree-layout#layout-tools), but the recommendations should be updated to depend on whether the executables are needed at runtime. Tools that applications will be using at runtime should be installed in both release and debug versions.

I need dbus-daemon and dbus-launch to work both from their location in vcpkg_installed/ and when copied to an install prefix together with other dependencies and my own software and I need this to work for both release and debug builds that use dynamic linking. My install prefix keeps both libraries and binaries in the same folder on Windows and on Linux my project uses the typical bin/ and lib/ folder. The requirements for this to work is that I have both the release and debug versions of the executables and on linux I would very much like the RUNPATH to include $ORIGIN/../lib, although I can patch that with `patchelf` when I install the executables to my CMAKE_INSTALL_PREFIX.

### Proposed solution

Update the dbus port to also install debug versions of all tools, but at least dbus-daemon and dbus-launch.

### Describe alternatives you've considered

If the software needs to be able to be able to run from an install prefix then the only alternatives I can think of is to use a system provided version of dbus-launch and dbus-daemon. This is not great and also does not work on all platforms.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tsondergaard**
dbus-daemon, dbus-launch and the other executables in the dbus port are integral components of the D-Bus system. libdbus-1 is of no utility in itself without dbus-daemon. A case could be made that libdbus-1 in Vcpkg could be used on Linux against the distro-provided dbus-daemon, but this is not true on Windows and other platforms where the dbus package is also available (and usefully so). Therefore these executables must be treated as first-class artifacts in the vcpkg port that can be installed along with shared libraries to a prefix by consuming applications.

The release version of the dbus-* executables are available in the tools folder, but that leaves a hole for dynamically linked debug builds. These cannot install the dbus-* executables from the tools folder as they are release binaries that are not (necessarily) compatible with the installed libdbus-1.so/dll and other shared library dependencies. To properly support debug builds with dynamic linking the debug variant of these executables must also be preserved. They cannot be moved to tools/<port>/debug as that change of directory layout is not compatible with the RUNPATH in the executables [$ORIGIN:$ORIGIN/../../lib]. Instead all dbus-* executables are left in the bin and bin/debug layout.

Fixes #37625

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by tsondergaard** [comment]
CI fails with "warning: The following EXEs were found in /bin or /debug/bin. EXEs are not valid distribution targets." I understand the drive not to become a general package manager, but libdbus is not of much use without dbus-daemon.

**Comment by dg0yt** [comment]
Executables must be moved to `tools` with `vcpkg_copy_tools(...)`.

File: ports/dbus/portfile.cmake [hunk 1]
```diff
@@ -74,15 +74,10 @@ vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/share/dbus-1/session.conf" "<inclu
...
-set(TOOLS daemon launch monitor run-session send test-tool update-activation-environment)
 if(VCPKG_TARGET_IS_WINDOWS)
     file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/tools/${PORT}")
     file(RENAME "${CURRENT_PACKAGES_DIR}/bin/dbus-env.bat" "${CURRENT_PACKAGES_DIR}/tools/${PORT}/dbus-env.bat")
     vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/tools/${PORT}/dbus-env.bat" "${CURRENT_PACKAGES_DIR}" "%~dp0/../..")
-else()
-    list(APPEND TOOLS cleanup-sockets uuidgen)
 endif()
-list(TRANSFORM TOOLS PREPEND "dbus-" )
-vcpkg_copy_tools(TOOL_NAMES ${TOOLS} AUTO_CLEAN)
```
Comment (dg0yt) [inline ports/dbus/portfile.cmake:86]: This cannot be removed.

**Comment by dg0yt** [comment]
Put the debug executables into `tools/${PORT}/debug`.

**Comment by dg0yt** [comment]
> They cannot be moved to tools//debug as that change of directory layout is not compatible with the RUNPATH in the executables [$ORIGIN:$ORIGIN/../../lib].

This must be handled correctly in `z_vcpkg_fixup_rpath.cmake`.

**Comment by tsondergaard** [comment]
> > They cannot be moved to tools//debug as that change of directory layout is not compatible with the RUNPATH in the executables [$ORIGIN:$ORIGIN/../../lib].
> 
> This must be handled correctly in `z_vcpkg_fixup_rpath.cmake`.

Would it be acceptable to install the debug executables to `debug/tools/<port>`? That would collect all debug content in the debug/ folder and that, very importantly, would give a consistent layout. This matters especially for dynamically linked code on linux with z_vcpkg_fixup_rpath.cmake.

**Comment by Neumann-A** [comment]
> change of directory layout is not compatible with the RUNPATH in the executables [$ORIGIN:$ORIGIN/../../lib]

Since nobody directly said it. Ignore the message the port prints after the portfile has executed. `z_vcpkg_fixup_rpath.cmake` will take care of it and adjust the RPATH accordingly if the location contains `debug/`.

**Comment by tsondergaard** [comment]
> Put the debug executables into `tools/${PORT}/debug`.

I've pushed an update that puts them in `debug/tools/${PORT}` rather than `tools/${PORT}/debug`. That way the debug layout is identical to the release layout. That makes a lot more sense to me. It also means that I my project can install the dbus-* executables directly without having to patch the RUNPATH.

**Comment by tsondergaard** [comment]
@dg0yt, why the "thumbs down" to installing in `debug/tools/${PORT}`? The CI system appears to accept it?

**Comment by dg0yt** [comment]
> @dg0yt, why the "thumbs down" to installing in `debug/tools/${PORT}`? The CI system appears to accept it?

CI accepts a lot which shouldn't go in... I could only give a quick reaction at that moment. But the right locations was already noted in previous comments. And the `debug` prefix isn't standalone anyways: no `share`, no  `etc`, no `include`. I don't say I like it that way.

**Comment by tsondergaard** [comment]
> I don't say I like it that way.

That’s how I feel, so I’m just going to see if I get a bit more engagement from vcpkg maintainers, before accepting a solution that requires me to use patchelf to fixup the executables when my project copies them to an install prefix

**Comment by tsondergaard** [comment]
I think it might be useful to elaborate on why I think `debug/tools/${PORT}` is much superior to `tools/${PORT}/debug`. First, it is consistent that all the debug items are in one folder that has the same layout as the release folder. That's just nice. Second is when an application wants to install the dependencies to CMAKE_INSTALL_PREFIX. This is what it will look like with `debug/tools/${PORT}` assuming the application uses the same `<prefix>/bin/`, `<prefix>/lib/` layout on Linux:

```
if (UNIX)
    set(MY_VCPKG_INSTALLED_DIR ${VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}$<$<CONFIG:Debug>:/debug>)
    install(PROGRAMS
            ${MY_VCPKG_INSTALLED_TOOLS_DIR}/dbus/dbus-daemon
            ${MY_VCPKG_INSTALLED_TOOLS_DIR}/dbus/dbus-launch
            DESTINATION bin)
endif()
```

if instead dbus-daemon and dbus-launch are installed in a different relative folder layout between the binary and the `lib/` directory, then a consuming application will need to fixup the RUNPATH on installation, which would look something like this:

```
if (UNIX)
    set(MY_VCPKG_INSTALLED_DIR ${VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET})
    install(CODE "
            execute_process(
                COMMAND install ${MY_VCPKG_INSTALLED_TOOLS_DIR}/dbus$<$<CONFIG:Debug>:/debug>/dbus-daemon ${CMAKE_INSTALL_PREFIX}/bin/dbus-daemon
            )
            execute_process(
                COMMAND patchelf --set-rpath $ORIGIN/../${CMAKE_INSTALL_LIBDIR} ${CMAKE_INSTALL_PREFIX}/bin/dbus-daemon
            )
            ")
    install(CODE "
            execute_process(
                COMMAND install ${MY_VCPKG_INSTALLED_TOOLS_DIR}/dbus$<$<CONFIG:Debug>:/debug>/dbus-launch ${CMAKE_INSTALL_PREFIX}/bin/dbus-launch
            )
            execute_process(
                COMMAND patchelf --set-rpath $ORIGIN/../${CMAKE_INSTALL_LIBDIR} ${CMAKE_INSTALL_PREFIX}/bin/dbus-launch
            )
           ")
endif()
```

As you can see this is way more clumsy.

**Comment by tsondergaard** [comment]
Oups, I think I’m wrong above. In my head I was think that the relative path from bin/ to lib/ was the same as tools/ to lib/, ie `$ORIGIN/../lib`, but with installation into `tools/<port>` the RUNPATH has to be `$ORIGIN/../../lib`. Drat. That means I need to `patchelf` the RUNPATH regardless. Only way to avoid patchelf is installation in a patch where RUNPATH can be `$ORIGIN/../lib`. That could be keep it in `bin/` which CI and established rules (strangely, imho) prohibits or install directly in `tools/` which, I believe is not encouraged either.

**Comment by data-queue** [comment]
I think debug tools should go in `debug/tools/${PORT}`. The other debug folders are `debug/lib` and `debug/bin` and everything debug would live in the same directory. @BillyONeal any thoughts?

**Comment by dg0yt** [comment]
> I think debug tools should go in `debug/tools/${PORT}`.

@data-queue These are the issues where your team is asked to clarify the installation layout for tools:
https://github.com/microsoft/vcpkg/issues/17607 (May 2021)
https://github.com/microsoft/vcpkg/issues/13649 (Sep 2020)

Answers to the open questions are not less urgent just because the issues are getting old...

**Comment by tsondergaard** [comment]
[This](https://github.com/microsoft/vcpkg/issues/13649#issuecomment-886536455) 3 year old comment on #13649 by @ras0219-msft says debug tools go in `debug/tools`. Is that still the policy? Next question is whether all executables are tools. `dbus-daemon` and `dbus-launch` are intended for runtime use, they are not build-time tools.

**Comment by BillyONeal** [comment]
I believe the status quo is that because vcpkg is explicitly not intended to be a *runtime* deployment system, we do not have a canonical location for these components.

> I understand the drive not to become a general package manager, but libdbus is not of much use without dbus-daemon.

I believe that means that the status quo is that we don't have a means of delivering libdbus in a way consistent with what libdbus customers want.

However, it also seems like these dbus components are of the kind where someone trying to provide them from 'outside the system' like this are always going to result in pain, since the version you select in vcpkg is unlikely to match versions on the system.

> but this is not true on Windows and other platforms where the dbus package is also available (and usefully so)

What happens if 2 windows programs do this and choose totally different versions of dbus?

**Comment by dg0yt** [comment]
FTR another per-config vcpkg port particularity which may interfer with deployments.
~~~
NOTE: system bus listens on /<VCPKG_ROOT>/packages/dbus_x64-linux-dynamic/debug/var/run/dbus/system_bus_socket
| This build of dbus will not interoperate with the well-known
| system bus socket, /var/run/dbus/system_bus_socket.
~~~

**Comment by tsondergaard** [comment]
> FTR another per-config vcpkg port particularity which may interfer with deployments.
> 
> ```
> NOTE: system bus listens on /<VCPKG_ROOT>/packages/dbus_x64-linux-dynamic/debug/var/run/dbus/system_bus_socket
> | This build of dbus will not interoperate with the well-known
> | system bus socket, /var/run/dbus/system_bus_socket.
> ```

This is a separate issue and only relevant if you will be using the system bus. I'm only using session busses.

**Comment by tsondergaard** [comment]
> I believe the status quo is that because vcpkg is explicitly not intended to be a _runtime_ deployment system, we do not have a canonical location for these components.
> 
> > I understand the drive not to become a general package manager, but libdbus is not of much use without dbus-daemon.
> 
> I believe that means that the status quo is that we don't have a means of delivering libdbus in a way consistent with what libdbus customers want.
> 
> However, it also seems like these dbus components are of the kind where someone trying to provide them from 'outside the system' like this are always going to result in pain, since the version you select in vcpkg is unlikely to match versions on the system.

This PR settles just installs the dbus-* executables in debug/tools/dbus. That should be okay, should it not? See https://github.com/microsoft/vcpkg/pull/37736#issuecomment-2026900349
 
> > but this is not true on Windows and other platforms where the dbus package is also available (and usefully so)
> 
> What happens if 2 windows programs do this and choose totally different versions of dbus?

dbus is very uncommon on Windows. If you use autolaunch: they would (try to) use the same bus. I don't know what compatibility guarantees libdbus has.

**Comment by dg0yt** [comment]
> > They cannot be moved to tools//debug as that change of directory layout is not compatible with the RUNPATH in the executables [$ORIGIN:$ORIGIN/../../lib].
> 
> This must be handled correctly in `z_vcpkg_fixup_rpath.cmake`.

`z_vcpkg_fixup_rpath.cmake` seem to be incorrect: It determines the debug type by matching the executable's directory against `debug/`. This works for 
- `debug/bin` (not allowed)
- `debug/libs`
- `debug/plugins/...`
- `debug/tools/...` (12 matches in x64-linux CI)
- `tools/.../debug/bin` (210 matches in x64-linux CI)

but not for
- `tools/.../debug` (a few matches in vcpkg CI).

**Comment by dg0yt** [comment]
> but not for
> 
>  * `tools/.../debug` (a few matches in vcpkg CI).

https://github.com/microsoft/vcpkg/pull/37964

**Comment by data-queue** [comment]
From the vcpkg team:
- The debug tools for `dbus` should be copied over to `debug/tools`.
- We are not saying that `debug/tools` is the accepted location for debug tools for all ports. They might need to go in other locations such as `tools/.../debug/bin`. And debug tools, in the general case, should not be copied over.

We are accepting this PR, and I will leave it open for a day or so before merging.

**Review by data-queue — APPROVED** [review]

**Review by JonLiu1993 — APPROVED** [review]

**Comment by dg0yt** [comment]
> From the vcpkg team:
> 
>  * The debug tools for `dbus` should be copied over to `debug/tools`.
> 
>  * We are not saying that `debug/tools` is the accepted location for debug tools for all ports. They might need to go in other locations such as `tools/.../debug/bin`. 

IMO this is a disappointing outcome. No reliable guideline for anything but `dbus`/now. No convergence. No perspective. In fact, little but our fading memories holding us back from moving the debug tools to another location in the next PR.

**Comment by dg0yt** [comment]
Reminder, the location was proposed due to workaround a defect in `z_vcpkg_fixup_rpath`. A fix is available.

**Comment by dg0yt** [comment]
Congrats. `dbus` won a free entry in the hall of shame at https://github.com/microsoft/vcpkg/issues/17607.

**Pull request merged**